src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/Continuations/TaskContinuationGenerator.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Reflection.Emit;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Datadog.Trace.ClrProfiler.CallTarget.Handlers.Continuations
+{
+    internal class TaskContinuationGenerator<TIntegration, TTarget, TReturn> : ContinuationGenerator<TIntegration, TTarget, TReturn>
+    {
+        private static readonly Func<TTarget, object, Exception, CallTargetState, object> _continuation;
+
+        static TaskContinuationGenerator()
+        {
+            DynamicMethod continuationMethod = IntegrationMapper.CreateAsyncEndMethodDelegate(typeof(TIntegration), typeof(TTarget), typeof(object));
+            if (continuationMethod != null)
+            {
+                _continuation = (Func<TTarget, object, Exception, CallTargetState, object>)continuationMethod.CreateDelegate(typeof(Func<TTarget, object, Exception, CallTargetState, object>));
+            }
+        }
+
+        public override TReturn SetContinuation(TTarget instance, TReturn returnValue, Exception exception, CallTargetState state)
+        {
+            if (_continuation == null)
+            {
+                return returnValue;
+            }
+
+            if (exception != null || returnValue == null)
+            {
+                _continuation(instance, default, exception, state);
+                return returnValue;
+            }
+
+            Task previousTask = FromTReturn<Task>(returnValue);
+            if (previousTask.Status == TaskStatus.RanToCompletion)
+            {
+                _continuation(instance, default, null, state);
+#if NET45
+                    // "If the supplied array/enumerable contains no tasks, the returned task will immediately transition to a RanToCompletion state before it's returned to the caller."
+                    // https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenall?redirectedfrom=MSDN&view=netframework-4.5.2#System_Threading_Tasks_Task_WhenAll_System_Threading_Tasks_Task___
+                    return ToTReturn(Task.WhenAll());
```

Comment:
Would it be better to cache a completed task like `Task.FromResult(false)` and return that instead of `Task.WhenAll()`? I'm concerned about `Task.WhenAll()` allocating a new object every time. Or could we actually return `previousTask`?

Reply:
I like that idea, we just can return `returnValue`. I'll do that.

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/1007/commits/aabd3b6c5bff6ddc3f4bf390646aa9184468e5ee